### ISSUE TITLE
question about build.sh

### ISSUE THREAD

**Issue by sunyi000** [singularityhub/circle-ci#4]
Hi, 

Sorry if I posted this question to the wrong place.

I tried to use the build.sh and the modified version of config.yml to push container image (.simg) files to my local singularity registry, but line 162 in that build.sh script caused an 'sregistry command not found' error. 

The command with --uri and --cli works fine without circleci. (if i install sregistry to a VM and run sregistry push manually)

Could anyone please provide some pointers on this issue?

Thank you very much

**Comment by vsoch** [comment]
Heyo! 

sregistry is installed with pip in setup.sh, which is run first in the config.yml. Did you run it to install sregistry and singularity? The intended use case for the scripts here are also to be run during the CI, and it sounds like you are using them outside of that case, FYI.

**Comment by sunyi000** [comment]
Thank you very much for your help.
I did run setup.sh in config.yml. what I did are

- I copied the whole .circleci folder into my bitbucket git repo
- linked my bitbucket repo with circleci
- created ENVIRONEMNT variables using circleci UI, including SREGISTRY_CLIENT and SREGISTRY_CLIENT_SECRETS
- copied the credential JSON file from my local SResigtry
- I replaced the line in config.yml
`/bin/bash ~/repo/.circleci/build.sh Singularity`
with
`/bin/bash ~/repo/.circleci/build.sh --uri collection/container --cli registry Singularity`

everything else in config.yml remain the same.

When I git push to remote bitbucket, it automatically triggered the circleci build. From the output in circleci UI, it installs sregistry and singularity. 

the first error I have is . venv/bin/activate No such file or directory
I then added `virtualenv venv` before that line to make the error go away. 
but now I got the error saying sregistry command not found in line 162 in build.sh 

I'm a bit confused about what are the causes..Also, I copied the json credential file to $HOME/.sregistry and specified this as an environment variable via circleci UI. Is this the correct way of doing it?

Thank you very much and really appreciate your help

**Comment by vsoch** [comment]
You are totally correct that there is a bug! I actually saw this when I was making the static registry and forgot to fix up the config here. Give me a few minutes and I'll get a working version up for you!

**Comment by vsoch** [comment]
okay I've fixed the venv error - copy the files again and give it a try! For the sregistry is not found, I added a "which sregistry" to the build step and it look like it's in `/opt/circleci/.pyenv/shims/sregistry`. So if you get that it's not found, let's try adding this line to the beginning of the block (right before you build)

```
export PATH=$PATH:/opt/circleci/.pyenv/shims
```
If that works let me know and I'll add to the example here!

**Comment by vsoch** [comment]
I added it, and just logged into the circle shell and can confirm it's found on the path:

```bash
$ which sregistry
/opt/circleci/.pyenv/shims/sregistry
```

**Comment by sunyi000** [comment]
Thank you Vanessa both error are fixed for me now. 
My sregistry path is /home/circleci/conda/registry so I exported this before build

another question, how do I tell circlecit my JSON credential file?
It now can't find the URL 
No schema supplied. Perhaps you meant http://None/upload/chunked_upload

It seems that it can't find the .sregistry file which contains the JSON credentials..

Thanks

**Comment by vsoch** [comment]
How are you adding the file? I would suggest not adding the file (this seems harder) but instead exporting the variables that you need as base64 encoded project secrets, and then decoding them when you export at runtime. There is a good example here: https://circleci.com/docs/2.0/env-vars/#setting-an-environment-variable-in-a-project

**Comment by vsoch** [comment]
I'm working on an example now (for a different storage, but similar with environment variables) so I will have some pointers for you definitely within the next few days, if we don't figure it out just chatting here.

**Comment by sunyi000** [comment]
Thanks
I tried to make that file in setup.sh and echo SREGISTRY_CLIENT_SECRET with the file path as the value in build.sh , but this obviously failed.

I will have a look at the example link and see if I can get it work.

Thank you very much

**Comment by vsoch** [comment]
Ah! So the idea with Sregistry client is that you can export the values to go into the file, and it will be created for you. Give that a go!

**Comment by sunyi000** [comment]
OK. I will give it a go and let you know 

Thank you

**Comment by sunyi000** [comment]
Hi Vanessa, 
I got it working this way

I created SREGISTRY_TOKEN, SREGISTRY_USERNAME and SREGISTRY_BASE via CircleCI UI settings for the project.

Then I modified build.sh to add the below line before build
`echo "{\"SREGISTRY_CLIENT\": \"registry\",\"registry\": {\"token\": \"$SREGISTRY_TOKEN\",\"username\": \"$SREGISTRY_USERNAME\",\"base\": \"$SREGISTRY_BASE\"}}" > $SREGISTRY_CLIENT_SECRETS`

and added environment in config.yml
`environment:
      SREGISTRY_CLIENT_SECRETS: /tmp/.sregistry
`

This works but all images from the git repository will be pushed to sregistry server under same username.

Is there a way to make circleci/build.sh read each user's own credential/environment variables? 

For example, if two developers in the same project commit to git repository, the above method will store two images under same account in my sregistry server. Is there a method to 'sregistry push' to different user accounts in sregistry server?

Thank you very much

**Comment by vsoch** [comment]
Hey @sunyi000 ! This is a very good question! We are working on similar builds in parallel, and I did something similar also for a credential file - adding instead the *entire* thing as a secure environment variable, echoing into a file, and then exporting the path as another environment variable, e.g.,:

```bash
source ${BASH_ENV}
echo $GOOGLE_APPLICATION_CREDENTIALS > /tmp/pancakes.json
export GOOGLE_APPLICATION_CREDENTIALS=/tmp/pancakes.json
```
See the file [here](https://github.com/singularityhub/busybox/blob/master/.circleci/config.yml#L95).

To answer your question - are you enabling user collections for your registry? Ideally you would use the Github repositories to manage permissions and containers. What I like to do is more of a lazy mans approach to managing the namespace, and have the Github username and reponame reflect the collection. For example, singularityhub/busybox goes into collection singularityhub/busybox, and you can do that automatically in the config.yml like:

```bash
NAMESPACE=${CIRCLE_PROJECT_USERNAME}/${CIRCLE_PROJECT_REPONAME}
echo "Container namespace is $NAMESPACE"
```
It follows that your users should manage separate GitHub repos per collection namespace, and then you would not need to have them share credentials or usernames.

So what I would do is to create a common configuration, and write up instructions for your users to export particular environment variables (or the whole file from their local machine), and instead of environment variables on the *project level* you would want them on the repository level. If you somehow need to share some credential you could put that on the project level, and then have the users change the repository names (collections) on the repo level. Do you want to share with me what you have done so far? I'd be happy to do this write up and add to the repo here, it's a good example to have since we are having this discussion!

**Comment by sunyi000** [comment]
Hi Vanessa,

Thank you very much for the detailed explanation.

We are a development team contains 8 developers working on different project within a University, and we use Singularity containers for some of the projects.

We use bitbucket repo for all the projects because we have an academic license so that we could have private repo, team users control etc for free.
 
CircleCi is our CI tool used across the team.

The purpose is to make circleci work for singularity container recipes files, and store the container images in a private hub so everyone in our team could share or reuse the container images.

Usually we will have 2-3 developers working on the same projects.

So to share why and what I have done so far.

I followed your instruction to set  up a sregistryhub on openstack research cloud. Basically just launch a instance and run the prepapre_server.sh and changed the settings. I enabled Github login to that server and made myself a superuser, because I remember in the documentation you mentioned that Github is the easiest to setup. I enabled user collections for our registry.

I have also created an Openstack object store and mounted it to the server using svfs to store container images. I have changed the /var/www/images in docker-compose mapping it to the mounted object store folder.

Then I copied the .circleci folder from this github repository to one of our Singularity recipe repository and tried to make it build automatically and upload the image. To make this work, 

1. I linked my bitbucket repo with circleci
2. modified config.yml added --uri and --cli to the build command
3. created environment variables via CircleCI UI (as mentioned in previouse comment)
4. added environment SREGISTRY_CLIENT_SECRETS to config.yml and 
5. copied the JSON credential from SRegistry server
6. echo the JSON content to build.sh before the build command, replacing token, username and base with environment variables created in circleci UI

What am currently looking are
- following your instruction in the documentation to add HTTPS to the server
- exploring the possibility to enable swift client so container images could be pushed to Openstack Object Storage directly. 
- log in to sregistry server using bitbucket account, although I'm not sure if this is possible.

Also, forgot to mention,I think there maybe another bug somewhere in the code. To specify parameters to the build.sh script I have to use -uri instead of --uri

Thank you again for your help
Yi

**Comment by vsoch** [comment]
Ah! I think I know what you need - I have an example circleci config that is intended to build (and deploy) to a static registry - it generates the container / collection names based on the path in the repository (and not the repository name itself) so what you would want to do is create the (project level) environment variable for the credential, but then push to sregistry as the same user (but your developers wouldn't get access to this credential) and the collection / namespace would be based on the folder hierarchy.

There is a bug with the -uri variable, I fixed it in other versions (but not the one you are going from!) 

Let me throw together a quick example for a circle build to sregistry that would honor the folder structure for collections.

**Comment by vsoch** [comment]
Hey working on this now! I'll see if I can get something for you to test soon :) It's super cool you have a small team and are deploying an sregistry and it's working well thus far! I'll do my best to help you finish up the details :)

**Comment by vsoch** [comment]
Almost done! I'm connecting to CircleCI now and testing. Adding bitbucket as a login is very reasonable, could you please open a new issue and I'll get to it when I finish this example for you?

**Comment by vsoch** [comment]
hey @sunyi000 great news! I just finished your example - it will work to build a set of containers (namespace based on the subfolders in the repository). There is a repository you can clone and then follow instructions in the README here -> https://github.com/singularityhub/circle-ci-sregistry.

I tested it through all steps minus the final push! Would you be able to follow the instructions, and test it out? When it works for you, I'd like to do a small write up to share with the Singularity community! If it's okay with you, I'd like to add your specific deployment / use case so others know about how you went about it.

I'm also just about finished up with the bitbucket integration - I'll open a PR to sregistry and ask you to set up your bitbucket to test it out. I'll ping you in the PR.

**Comment by sunyi000** [comment]
Hi Vanessa,
Thank you for the example. I'm happy to test everything and share our use case. 

I'm away from work today so I will test them and feedback to you next Monday.

Thank you
Yi

**Comment by vsoch** [comment]
@sunyi000 the pull request to add Bitbucket is here -> https://github.com/singularityhub/sregistry/pull/173 Please read the description and test the branch, and let me know if you run into any issues. When you've confirmed all is working as expected we can merge, and the container will build to latest along with the commit (as tag) for the repository to Docker Hub.

**Comment by vsoch** [comment]
We posted at (almost) the same time! Got it @sunyi000 - have a good weekend and chat next week!

**Comment by sunyi000** [comment]
Hi @vsoch ,

a little bit updates on my testing. 

just two minor things i noticed,
1. in the Prepare Sregistry Environment step, I need to change 
cat "${SREGISTRY_CREDENTIALS}" > $HOME/.sregistry
to 
echo "${SREGISTRY_CREDENTIALS}" > $HOME/.sregistry 

2. I got the  'sregistry command not found ' error again when execute the final push. 
I guess I need to export the correct sregistry PATH, I'm doing it now..

3. I haven't tested the bitbucket one yet.. I will update again soon.

Thank you very much
Yi

**Comment by sunyi000** [comment]
the push worked after I export /opt/circleci/.pyenv/shims/ in the PATH.

I will test the bitbucket PR soon.

**Comment by vsoch** [comment]
Okay, I did the fix for the first item, testing the second!

**Comment by vsoch** [comment]
Just merged changes -> https://github.com/singularityhub/circle-ci-sregistry/pull/2 the cat is changed to an echo, and sregistry is confirmed to exist in the final step.

**Comment by sunyi000** [comment]
Hi @vsoch ,

it worked for me. Thanks very much.

For the bitbucket PR, I have merged the changes and modified config.py, secrets.py etc..but i have got errors start  the  containers. the error is at shub/apps/logs/models.. error trace is as below

`
Traceback (most recent call last):
   File "shub/wsgi.py", line 7, in <module>
      application = get_wsgi_application()
   File "/usr/local/lib/python3.5/site-packages/django/core/wsgi.py", line 12, in get_wsgi_application
      django.setup(set_prefix=False)
   File "/usr/local/lib/python3.5/site-packages/django/__init__.py", line 24, in setup
      apps.populate(settings.INSTALLED_APPS)
    File "/usr/local/lib/python3.5/site-packages/django/apps/registry.py", line 112, in populate
      app_config.import_models()
   File "/usr/local/lib/python3.5/site-packages/django/apps/config.py", line 198, in import_models
      self.models_module = import_module(models_module_name)
   File "/usr/local/lib/python3.5/importlib/__init__.py", line 126, in import_module
      return _bootstrap._gcd_import(name[level:], package, level)
   File "<frozen importlib._bootstrap>", line 986, in _gcd_import
   File "<frozen importlib._bootstrap>", line 969, in _find_and_load
   File "<frozen importlib._bootstrap>", line 958, in _find_and_load_unlocked
   File "<frozen importlib._bootstrap>", line 673, in _load_unlocked
   File "<frozen importlib._bootstrap_external>", line 662, in exec_module
   File "<frozen importlib._bootstrap>", line 222, in _call_with_frames_removed
   File "./shub/apps/logs/models.py", line 16, in <module>
      class BaseAPIRequestLog(models.Model):
   File "./shub/apps/logs/models.py", line 18, in BaseAPIRequestLog
     user = models.ForeignKey(settings.AUTH_USER_MODEL, null=True, blank=True)
 TypeError: __init__() missing 1 required positional argument: 'on_delete'
Mon Feb 11 22:36:51 2019 - unable to load app 0 (mountpoint='') (callable not found or import error)
Mon Feb 11 22:36:51 2019 - *** no app loaded. going in full dynamic mode ***
Mon Feb 11 22:36:51 2019 - *** WARNING: you are running uWSGI as root !!! (use the --uid flag) *** 
Mon Feb 11 22:36:51 2019 - *** uWSGI is running in multiple interpreter mode ***
Mon Feb 11 22:36:51 2019 - spawned uWSGI master process (pid: 50)
Mon Feb 11 22:36:51 2019 - spawned uWSGI worker 1 (pid: 54, cores: 4)
Mon Feb 11 22:36:51 2019 - spawned uWSGI worker 2 (pid: 58, cores: 4)
Mon Feb 11 22:36:51 2019 - spawned uWSGI worker 3 (pid: 62, cores: 4)
 Mon Feb 11 22:36:51 2019 - spawned uWSGI worker 4 (pid: 66, cores: 4)
Mon Feb 11 22:37:10 2019 - --- no python application found, check your startup logs for errors ---

`

**Comment by vsoch** [comment]
hey @sunyi000 ! This is a bug I missed when updating the django version:

https://www.valentinog.com/blog/django-missing-argument-on-delete/

I'll get in a PR later today to fix it, and let me know what you do if you wind up trying the fix first!

**Comment by vsoch** [comment]
Are you able to give me enough to reproduce the error? I just cloned and did docker-compose up -d (adding a SECRET_KEY to settings) and I can't reproduce.

**Comment by sunyi000** [comment]
Hi Venessa,
This is what I did,
1. clone the repository to /opt/sregistry
2. merged the bitbucket changes. (I did this by git merge)
3. added POSTGRES_PASSWORD and POSTGRES_USER as environment variables to db images in docker-compose.yml. Maped the /var/www/images to somewhere in /mnt
4. enabled bitbucket and github oauth2 in config.py
5. added db config json and github, bitbucket client id and secrets into secrets.py, together with other changes
6. commented out the last  if- then statement in prepare_server.sh
7. run prepare_server.sh
8. cd to /opt and run  docker build
9. then docker-compose up -d

after docker-compose only uwsgi, ngix,redis and postgres containers started. worker container not started. so I did a docker-compose logs and saw the above error trace...

Yi

**Comment by vsoch** [comment]
For a sanity check can you please try without using a custom database? Meaning, just use the default Postgres that comes as a container.

**Comment by vsoch** [comment]
And it’s not clear to me what you mean by merged bitbucket changes - why is sregistry relevant to bitbucket?

**Comment by sunyi000** [comment]
I did a git clone https://github.com/singularityhub/sregistry.git and switched to add/bitbucket branch, before docker build and docker-compose up

**Comment by vsoch** [comment]
Ah ok! I’ll need to pick this up tomorrow it’s 3:30 am here, will ping you if/when I can reproduce!

**Comment by sunyi000** [comment]
I will try the default postgres db

**Comment by sunyi000** [comment]
not a problem..time for me to finish work too..
talk tomorrow

**Comment by vsoch** [comment]
okay just reproduced! Will get a fix for you asap.

**Comment by vsoch** [comment]
okey doke! The PR is updated to fix the issues. Apologies for this oversight - Django was updated to 2.0 and there were some things that I missed.

**Comment by sunyi000** [comment]
Hi @vsoch ,
still error in uwsg_1 container about slqalchemy. the error comes from 
/shub/apps/users/views/auth.py

WARNING Database disabled. Install sqlalchemy for full functionality
uwsgi_1   | Internal Server Error: /login/
uwsgi_q   | AttributeError: 'WSGIRequest' object has no attribute 'user'

this is happening when clicking 'Login' from the registryhub

Did I miss anything?

Thanks very much

**Comment by vsoch** [comment]
heyo @sunyi000 ! Thanks for reporting this issue - it looks like Django 2.0 updated the request.user object so that `request.user.is_authenticated` is now a boolean (previously it was a function, called like `request.user.is_authenticated().` The specific error you found was also another change - the MIDDLEWARE classes in settings used to be under the variable `MIDDLEWARE_CLASSES` and 2.0 updated them to just `MIDDLEWARE`. I was able to reproduce the error, fix it, and tested logging in with a local GitHub, so I think it's ready for you to give another go! 

For the sqlalchemy bit - disregard! The sregistry client has an installation option to install without the small local database (which is sqlite3) so this warning is just referring to that.

**Comment by sunyi000** [comment]
Hi @vsoch 
I have pulled your latest changes and my registryhub is up and running without errors..: )

just to clarify the bitbucket callback url? I used /complete/bitbucket but doesn't seem to be right..

Thanks

**Comment by vsoch** [comment]
We're getting close! Calllback URLS are very fussy in that they must be *exact.* If you are still testing on localhost, you will want to do something like `http://127.0.0.1/complete/bitbucket` Once you have a domain that changes to `http://<domain>/complete/bitbucket`, and once you have https then you also have that `https://<domain>/complete/bitbucket`. It has to match to a T - so try with and without ending slashes too.

**Comment by sunyi000** [comment]
Thank you @vsoch 

I still couldnt login using bitbucket. I have checked my consume key,secret,scope and tested callback url with and without the /

I found the below error in docker logs when building uwsgi container.. am not certain if that's related?

  File "/usr/local/lib/python3.5/site-packages/django/db/migrations/graph.py", line 104, in raise_error
    raise NodeNotFoundError(self.error_message, self.key, origin=self.origin)
django.db.migrations.exceptions.NodeNotFoundError: Migration users.0001_initial dependencies reference nonexistent parent node ('auth', '0009_alter_user_last_name_max_length')

**Comment by sunyi000** [comment]
Updates:

I did a clean rebuild of all images, and then run docker-compose up, the above error is gone.

but still not able to login via bitbucket (log in using github works). The only thing in the uwsgi container log is 'Internal Server Error' after click 'Login with Bitbucket '

In the docker-compose log, before I click anything on the registryhub, there are some db connection refused errors in it. it seems the uwsgi container trying to connect to posgres before the posgres is created. Is this the cause?

I will continue to see if I can find anything else
Thanks

**Comment by sunyi000** [comment]
Hi @vsoch ,

I can't push to sregistryhub any more..

1.when login with Github, and clicking 'Collections', page show 500 Internal server error.
  File "./shub/apps/users/views/auth.py", line 49, in validate_credentials
    if not user.is_anonymous():
TypeError: 'bool' object is not callable

2. if I run sregistry push command manually, error is
 File "./shub/apps/api/actions/push.py", line 61, in collection_auth_check
    raise PermissionDenied(detail="Unauthorized")
rest_framework.exceptions.PermissionDenied: Unauthorized

I have tried by setting my enviroment variables SREGISTRY_CLIENT_SECRETS to .sregistry file ,as well as SREGISTRY_CREDENDTIALS equal to the json content of .sregistry file.

can you please help .
thanks,
Yi

**Comment by sunyi000** [comment]
Hi @vsoch ,

I can't push to sregistryhub any more..

1.when login with Github, and clicking 'Containers', page show 500 Internal server error.
File "./shub/apps/users/views/auth.py", line 49, in validate_credentials
if not user.is_anonymous():
TypeError: 'bool' object is not callable

 2.  if I run sregistry push command manually, error is
       File "./shub/apps/api/actions/push.py", line 61, in collection_auth_check
       raise PermissionDenied(detail="Unauthorized")
       rest_framework.exceptions.PermissionDenied: Unauthorized


I have tried by setting my enviroment variables SREGISTRY_CLIENT_SECRETS to .sregistry file ,as well as SREGISTRY_CREDENDTIALS equal to the json content of .sregistry file.

I can see 'admin' stuff and django admin page after login with github account

can you please help .
thanks,
Yi

**Comment by vsoch** [comment]
Ack, another bug that I missed! I just fixed the "bool is not callable issue" - can you try again?

**Comment by sunyi000** [comment]
Thanks, it works this time..the whole thing works with Github credentials

still trying to troubleshoot Bitbucket Login..

**Comment by vsoch** [comment]
I think it would be best for me to give this a try now - maybe not today but definitely soon. Will let you know if I have luck! I can definitely say from experience the call back urls are quite fussy, and some don't even allow http.

**Comment by vsoch** [comment]
okay, I'm trying to reproduce, and hitting that the backend isn't found:

![image](https://user-images.githubusercontent.com/814322/52980640-1e28f100-33a9-11e9-8947-92ad6c8c5c7c.png)

**Comment by vsoch** [comment]
https://github.com/python-social-auth/social-core/issues/318

**Comment by sunyi000** [comment]
Hi Vanessa,

a quick question about copying files into container using %files

what is the best way to do so in this circleci build environment?

I currently must put the full absolute circleci repo path under %files in the recipe file for it to work. 
for example

```
%files
/home/circleci/repo/container/test/config.ini /tmp
```
relative path such as ./config.ini or just config.ini won't work in the circleci build, but works without circleci

Thanks very much

**Comment by vsoch** [comment]
oh, this looks like a bug! Which template are you using?

**Comment by sunyi000** [comment]
Hello @vsoch ,

I'm using the template here
https://github.com/singularityhub/circle-ci-sregistry

I think the 'TypeError: 'bool' object is not callable' is still there. This time it doesn't happen when you click "Containers" from the menu, but it does appear when I create a collection after clicking 'Submit'.

I was redirected to an error page after clicking 'Submit', but the collection is actually created successfully.

To replicate, 

1.Click Containers from the top menu - > click 'New Collection' button -> enter collection name -> Submit -> redirected to error page with 'TypeError' shown in log file
then click Browser Back button then the new collection was actually created.
 click the newly created collection name -> same error appear again

or

2.follow the docker log of the uwsgi container -> run 'sregistry push' from another machine 
an "Unauthorized" error will appear in docker log, despite all credentials and env variables are correct. I guess this maybe related to the bool object not callable error?

   raise PermissionDenied(detail="Unauthorized")
rest_framework.exceptions.PermissionDenied: Unauthorized


I'm using the 483cf90bc5dc3a985b9a11885df48c8489a392f9 commit for sregistryhub.

PS: this thread is getting longer, do you mind if I keeping adding questions here? :)

thank you

**Comment by sunyi000** [comment]
OK, I modified models.py file, and changed line 267, and the issue 1,'bool not callable' issue, mentioned above goes away...
is_authenticated()
to
is_authenticated

but the second permission issue is still there...it complains me is not a valid user, request invalid when calling
/api/upload/chunked_upload

**Comment by vsoch** [comment]
did I not modify /fix this for you? You shouldn't be modifying the source files, otherwise I can't fix bugs in the source code and have you test :/

**Comment by vsoch** [comment]
I fixed the is_authenticated issues, could you pull and test again, and please give me enough to reproduce the nginx upload error?

**Comment by vsoch** [comment]
okay I fixed the issue in https://github.com/singularityhub/circle-ci-sregistry. For the future it would be helpful if you could report issues on the repositories where they are relevant - it gets confusing to have all these reported here for all different repos!

**Comment by sunyi000** [comment]
That bool error initially occurred when clicking the collection top menu, I think you fixed it therefore I can go to the next page but it happens again when 'submit' the collection.

So am just trying to troubleshoot locally and changing the file locally.

I thought the other repo circle-ci-singularity is an example you created based on my use case I reported in this thread. They are the same anyway I will post issues in the right repo in the future.

I will post issues in the right repos then..

**Comment by vsoch** [comment]
hey I'm here to help you! Can you please pull the updated branch, and then report the issue again when you push (the permission denied?) I'd like to focus on one error, debug it, and then get your feedback if it's resolved. We can do this together, and hey in retrospect no worries about where you post, I'll figure it out :)

**Comment by vsoch** [comment]
hey @sunyi000 , just to let you know I updated the recipe here to use the CircleCI orb. If you have any further issues with your registry, please open at https://www.github.com/singularityhub/sregistry. I'm going to close the issue here, since the original was resolved. Thanks!
